src/Icicle/Source/Checker/Constraint.hs

```diff
@@ -518,39 +537,38 @@ generateP
   -> Type n                 -- ^ result temporality
   -> [(Pattern n, Exp a n)] -- ^ pattern and alternative
   -> GenEnv n
-  -> GenConstraintSet a n
   -> Gen a n ([((Pattern n, Exp'C a n), SubstT n)], GenConstraintSet a n)
 
-generateP ann _ _ resTm [] _ cons
- = do   let cons' = require ann (CEquals resTm TemporalityPure) cons
+generateP ann _ _ resTm [] _
+ = do   let cons' = require ann (CEquals resTm TemporalityPure)
         return ([], cons')
 
-generateP ann scrutTy resTy resTm ((pat, alt):rest) env cons
+generateP ann scrutTy resTy resTm ((pat, alt):rest) env
  = do   (t, envp) <- goPat pat env
 
         let (_,_,datS) = decomposeT $ canonT scrutTy
-        let conss      = require (annotOfExp alt) (CEquals datS t) cons
+        let conss      = require (annotOfExp alt) (CEquals datS t)
 
-        (alt', sub, consa) <- generateX alt envp conss
+        (alt', sub, consa) <- generateX alt envp
 
-        let altTy'  = annResult (annotOfExp alt')
-        altTy      <- TypeVar <$> fresh
-        altTp      <- TypeVar <$> fresh
+        let altTy' = annResult (annotOfExp alt')
+        let altTp  = getTemporalityOrPure altTy'
```

Comment:
ok, I see. I tried something similar, but I think I tried to join temporalities on the whole type, rather than unwrapping first here. Anyway, this is good.